scripts/auto_archive.py

```diff
@@ -0,0 +1,202 @@
+
+    Requires to set GITHUB_ACCESS_TOKEN as an environment variable.
+
+    Parameters
+    ----------
+    since : Optional[datetime], optional
+        Start date from which commits are retrieved, by default date of the previous crawl, if never crawled set to
+        one week ago.
+    until : Optional[datetime], optional
+        Latest date at which commit are retrieved, by default `now`
+
+    Returns
+    -------
+    set[str]
+        Path of the dataset folders.
+    """
+    now = datetime.now().astimezone()
+
+    if since is None:
+        if os.path.exists(".conp-archive"):
+            with open(".conp-archive") as fin:
+                since = datetime.fromisoformat(fin.read())
+        else:
+            since = now - timedelta(weeks=1)
+
+    if until is None:
+        until = now
+
+    try:
+        gh_access_token = os.environ.get("GITHUB_ACCESS_TOKEN", None)
+        if gh_access_token is None:
+            raise OSError("GITHUB_ACCESS_TOKEN is not defined.")
+
+    except OSError as e:
+        # The program is not stopped since GitHub allows 60 query per hours with
+        # authentication. However the program will most likely fail.
+        logger.critical(e)
+
+    logger.info(f"Retrieving modfied datasets since {since}")
+    repo = Github(gh_access_token).get_repo("CONP-PCNO/conp-dataset")
+    commits = repo.get_commits(since=since, until=until)
+
+    with open(".conp-archive", "w") as fout:
+        fout.write(now.isoformat())
+
+    modified_datasets: set[str] = {
+        file_.filename
+        for commit in commits
+        for file_ in commit.files
+        if file_.filename.startswith("projects/")
+    }
+
+    return modified_datasets
+
+
+def archive_dataset(dataset_path: str, out_dir: str, version: str) -> None:
+    os.makedirs(os.path.dirname(out_dir), mode=0o755, exist_ok=True)
+    out_filename = f"{out_dir}_version-{version}.tar.gz"
+    logger.info(f"Archiving dataset: {dataset_path} to {out_filename}")
+
+    cwd = os.getcwd()
+    try:
+        datalad_archiver = export_archive.ExportArchive()
+        dataset_repo = git.Repo(dataset_path)
+
+        with open(os.path.join(dataset_path, ".git.log"), "w") as fout:
+            fout.write(dataset_repo.git.log(pretty="format:%H %s"))
+
+        # Export is performed from the dataset root.
+        # This is to avoid failure when a submodule is not downloaded; e.g. for parent
+        # dataset in dataset derivative.
+        os.chdir(os.path.join(cwd, dataset_path))
+        datalad_archiver(".", filename=out_filename)
+
+    except Exception as e:
+        raise ArchiveFailed(
+            f"FAILURE: could not archive dataset: {dataset_path} to {out_filename}\n{e}"
+        )
+    finally:
+        os.chdir(cwd)
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    datasets = get_all_datasets() if args.all else get_modified_datasets()
+
+    for dataset in datasets:
+        dataset_name = dataset.removeprefix("projects/")
```

Comment:
@mathdugre  this will not work for earlier versions of Python. That only exists for v3.9... I am running version 3.7 on my sandbox so the script is currently crashing for me.

Reply:
@cmadjar which Python version should we support for this tool? Currently, for the test suite we are only supporting 3.9+.